Add support for manually capturing charges in UPE

cesarcosta99 opened the pull request:
# Changes proposed in this Pull Request:
Resolves #1786 

This PR adds support for manually capturing charges in UPE checkout. The capture flag was being ignored (always set to `yes`) and not respecting the configuration that is saved in the settings about whether it should be automatic or manual.

# Testing instructions

Assuming you have the store up and running in test mode:

1. Ensure you have `_wcstripe_feature_upe` set to `yes` in `wp_options` table.
2. Visit the store and log in as admin.
3. In the [Stripe settings page](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe), uncheck `Capture charge immediately` option and save.
4. Go to the store > `Shop` and add a product to the cart.
5. Go to checkout and pay using a test card (e.g. `4242 4242 4242 4242`).
6. Back to the WP Admin, go to `WooCommerce` > `Orders` > the order you just created.
7. The order should have status `On hold` and a similar note:

<img src="https://user-images.githubusercontent.com/10233985/132069853-af033dc1-8c98-4f3c-8d38-387ca7d81beb.png" alt="screenshot" width="300" />

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

cesarcosta99 committed:
UPE: Add capture method param when creating intent

cesarcosta99 commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -461,7 +461,7 @@ public function process_upe_redirect_payment( $order_id, $intent_id, $save_payme
 			// }
 
-			$intent->captured = 'yes'; // TODO: this is to re-use the parent logic, maybe re-implement it to not use charges?, see: https://github.com/Automattic/woocommerce-payments/blob/b69532e92a381cf054f515896c56cc365e3904d4/includes/class-wc-payment-gateway-wcpay.php#L1125
+			$intent->captured = end( $intent->charges->data )->captured; // TODO: this is to re-use the parent logic, maybe re-implement it to not use charges?, see: https://github.com/Automattic/woocommerce-payments/blob/b69532e92a381cf054f515896c56cc365e3904d4/includes/class-wc-payment-gateway-wcpay.php#L1125
```
Will create an issue for this TODO. This refactor can be addressed in a separate PR as we would have to make a few changes in the _legacy_ checkout and rewrite part of the implementation in the abstract class.

cesarcosta99 replied:
Created #1790

superdav42 commented:
Ran into a small problem.
![image](https://user-images.githubusercontent.com/1534605/132704549-eb56f6a5-081f-4ffd-8f62-60b91c7c67a1.png)
Probably we need to add a property to the payment methods to indicate they support manual capture method. When capture method is set at manual only show the methods which support manual.

superdav42 commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -298,8 +298,10 @@ public function create_payment_intent_ajax() {
 	public function create_payment_intent( $order_id = null ) {
-		$amount = WC()->cart->get_total( false );
-		$order  = wc_get_order( $order_id );
+		$settings = get_option( 'woocommerce_stripe_settings', [] );
```
Really we should get the option from the gateway with `$gateway->get_option( 'capture' );` instead of accessing other options directly.

cesarcosta99 committed:
Refactor capture option retrieval

cesarcosta99 committed:
Fix available payment methods when manual capture is active

cesarcosta99 committed:
Fix lint error

cesarcosta99 commented:
Rebased `develop`.

cesarcosta99 commented:
> Ran into a small problem.
> ![image](https://user-images.githubusercontent.com/1534605/132704549-eb56f6a5-081f-4ffd-8f62-60b91c7c67a1.png)
> Probably we need to add a property to the payment methods to indicate they support manual capture method. When capture method is set at manual only show the methods which support manual.

Good catch, thanks! After rebasing `develop`, I realized unticking `Capture charge immediately` and saving, in the settings, will disable all these unsupported payment methods. However I also added an implementation to filter them out of the supported payment methods when creating the payment intent.

superdav42 commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -276,12 +276,11 @@ private function get_enabled_payment_method_config() {
-				'card',
-			]
-		);
+		$capture = ! empty( $this->get_option( 'capture' ) ) && $this->get_option( 'capture' ) === 'yes';
```
I think maybe this check should be moved to `get_upe_enabled_at_checkout_payment_method_ids()`. Which would still allow the payment methods to appear as enabled in the admin. It would require modifying WC_Stripe_UPE_Payment_Method::is_enabled_at_checkout(). 

It's fine to leave the check here too, I guess the real question is if card is the only payment methods that will ever support manual capture.

cesarcosta99 replied:
I think it's okay to have them disabled in the admin because that reflects what the customers see in the checkout. On the other hand, that can be confusing for the merchant that "will not be able" to enable other payment methods.

I'm not sure if that's covered in the new settings, but we can maybe add some sort of help note/tooltip in the option to warn the merchant that disabling the automatic capture will restrict the payment methods to card only or something in that direction.

superdav42 reviewed (approved):
It's good the way it is but maybe we want to refactor it further.

cesarcosta99 committed:
Default capture method to automatic when creating intent

cesarcosta99 committed:
Create tests for intent controller

At this point, there was no tests for the controller, now some tests are being added to cover the capture method parameter in the intent creation.

cesarcosta99 committed:
Fix variable name

cesarcosta99 committed:
Fix intent controller tests

Replace update option to avoid compatibility issues

cesarcosta99 committed:
Merge branch 'develop' into add/1786-manual-capture
